### ISSUE TITLE
Feature: automatically replace brackets with braces when contextually required

### ISSUE THREAD

**Issue by afetisov** [intellij-rust/intellij-rust#10397]
## Problem description

In a recent [IRLO discussion](https://internals.rust-lang.org/t/a-look-at-pros-and-cons-of-brackets-on-if-statements/18731), the following issue was raised:

> However, people still have to press `shift+[` and `shift+]` on every small if statement

and also

> Heh, I wish it were as ergonomic on all layouts as the US layout’s shift plus the keys to the right of P. For example, on Nordic (and eg. German) keyboards which have to accommodate three extra letters, {} are typed with altgr+7 and altgr+0, and on Mac even more awkwardly with alt+shift+8 and alt+shift+9! Looks like the French layout is even worse…

Indeed, typing braces can be slightly annoying on common keyboard layouts. I have thought about the following feature: what if a user could just type `[`, and the IDE would automatically change it into a `{`? This would obviously be not always desirable, because a user could have just wanted to type an array or subscript, so here are some more precise rules:

* In a context where `{` is syntactically valid  and `[` is not (after `else` or `loop`, after the function's return type, after `async` or `try`, after the name and generic parameters of a data type definition, after `mod foo`, etc)...
  * If a user types `[`, it is interpreted as typing `{`, including brace matching if applicable.

* In a context where `{` is syntactically expected, but `[` is possible as a part of the preceding expression (after a fully formed conditional in `if` expression or `while` block, after a fully formed `match` scrutinee, etc)...
  * If the user types `[` followed by newline (i.e. without trailing spaces or other following symbols)...
  * Then it is interpreted as typing `{`, and the typed bracket is replaced with a brace, together with its matching pair if present.

Examples:
```rust
fn foo(bar: Bar) /*caret*//*type [*/

// result
fn foo(bar: Bar) { /*caret*/ }
```

```rust 
fn foo(b: Bar) {
    if b [/*caret*//*type Enter*/
}

// result
fn foo(b: Bar) {
    if b { /*caret*/ }
}
```

Do you think it's a reasonable feature? 

Is there something like that already in the IDE? It looks like something which would be helpful in other languages than Rust. In Java or C, it looks even more useful, since they are not expression-based and there are far fewer places where a braced block can be used.

Are there any other cases where such symbol reinterpretation could be useful?
